### PR TITLE
Replace ThemeToggleButton with ThemeSelector

### LINKED ISSUE #3719: Feature Request: Add pre-defined themes user can select
**What would you like to be added**:

***(This is more than toggling Dark mode on and off)***

A drop-down to select a theme from a list of built-in themes. Selecting an option from the drop-down will render the application using the CSS styles associated with the chosen theme. The selected theme will be persisted so that when the user signs in again, they do not have to re-select the theme from the drop-down.

**Why would you like this to be added**:

An increasing amount of web applications and tools are now allowing users to customize the appearance by going to the "settings" and selecting a theme from a list. Some popular web apps/services include StackOverflow and GitHub.

GitHub offers some great customization options:

Screen Shot 2022-10-14 at 11 20 13 PM

### PR DESCRIPTION
**Pull request by Genne23v**
## Issue This PR Addresses
This PR addresses potentially fixes #3719 that provides additional themes for users to select. I have added `Light High Contrast` and `Dark Dim` colour themes in this PR. 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [x] **UI**: Change which improves UI

## Description
I have added `ThemeSelector` components as requested in issue #3719. I have removed toggle button and add popover selector to change the theme. 

Screenshot 2022-11-16 at 10 05 31 AM

There might be an update needed to support `GitHubContributionCard` but I don't see any part that imports `GitHubContributionCard` module. Please let me know if I miss anything. 

## Steps to test the PR
1. Scroll down on main page until the sidebar menu is displayed
2. Click the last palette icon
3. Select the theme to see if it's working
No special setup is required

## Checklist
I think this PR needs a UI test. But I don't see UI test set up in this project. Please let me know if any test needs to be written for my PR. 

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Review by manekenpix — CHANGES_REQUESTED** [review]
Looking great so far, just a few comments:

- Hovering over the theme selector button adds some colours that doesn't match what we do with other buttons.
image

- Would it be possible to use our palette for this menu?
image

- Switching from dark to light doesn't affect the theme selector menu
image

- Instead of having to click in the radio button, would it be possible to turn the whole "row" into a button. At first I was clicking on the whole thing thinking that'd do it, it took me a bit to figure out that I actually had to click on the right side of the row.

**Comment by Genne23v** [comment]
@manekenpix Thank you for the review. They are all relevant feedbacks. I will address them as quickly as possible.

**Review by cychu42 — CHANGES_REQUESTED** [review]
This looks great. I’m impressed by what you have done.
One thing I notice is that it feels hard to read some text in the high contract light theme. 
I know there are other theme options if one doesn't work well, but is this intentional?
image

Also, I notice the `Change Colour Theme` text in the theme menu acts as like a clickable button instead of simple text. You can click it, and it "flashes" like a button. I wonder if you can stop it from doing that.
image

**Comment by Genne23v** [comment]
@cychu42  Thanks for the feedback. Colour set is not intentional at all. I haven't been able to see with the actual content due to setup issue. I will arrange colours accordingly. 
And I will try to handle the tooltip as well.

**Comment by cychu42** [comment]
For the setup, try to ask for help on Slack if you need to. I managed to get it to run on Windows, but there are a lot of people having different issues, and I unfortunately don't know all of the answers.

**Comment by Genne23v** [comment]
I realized how careful @cychu42 @manekenpix reviewed my change. I think I addressed them all. Please let me know if you  find something with my updated branch. Thanks again.

**Review by cychu42 — CHANGES_REQUESTED** [review]
Nice! I see you fixed how the icon reacts to being clicked, and the pop-up menu changes colour with the theme.
However, I still find `Change Colour Theme` to act like it's clickable, and high contrast light theme still has the same colour scheme.
image
image
I wonder if we should save the design of colour combination to another follow-up issue. The basic of what the issue wanted is there already.

File: src/web/app/src/components/ThemeProvider.tsx [hunk 1]
```diff
@@ -1,19 +1,17 @@
...
 import { Theme } from '@material-ui/core/styles';
-
-import { ThemeName } from '../interfaces';
 import { lightTheme } from '../theme';
 
 type ThemeContextType = {
   theme: Theme;
-  themeName: ThemeName;
-  toggleTheme: () => void;
+  preferredTheme: String | undefined;
```
Comment (manekenpix) [inline src/web/app/src/components/ThemeProvider.tsx:7]: Why do we need this to have the option to be undefined?

File: src/web/app/src/components/ThemeSelector.tsx [hunk 2]
```diff
@@ -0,0 +1,133 @@
+// eslint-disable-next-line no-use-before-define
+import React, { useState } from 'react';
```
Comment (manekenpix) [inline src/web/app/src/components/ThemeSelector.tsx:2]:
```suggestion
import { useState } from 'react';
```

File: src/web/app/src/components/ThemeSelector.tsx [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  const handleClick = (e: React.MouseEvent<HTMLElement>) => {
+    setAnchorEl(e.currentTarget);
+  };
+
+  const handleClose = () => {
+    setAnchorEl(null);
+  };
+
+  const handleChange = (id: String): void => {
+    console.log('theme in ThemeSelector', preferredTheme, id);
```
Comment (manekenpix) [inline src/web/app/src/components/ThemeSelector.tsx:70]: I don't think we need to do this.

File: src/web/app/src/components/ThemeSelector.tsx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    changeTheme(id);
+  };
+
+  const open = Boolean(anchorEl);
+  const id = open ? 'simple-popover' : undefined;
```
Comment (manekenpix) [inline src/web/app/src/components/ThemeSelector.tsx:75]: Optionally, you can cast `anchorEl` to a boolean type using the double exclamation operator, like:

```ts
const id = !!anchorEl;
```

**Review by manekenpix — CHANGES_REQUESTED** [review]
Just a few comments:

- The light themes are using a colour defined in the `darkdimTheme` no matter which theme is selected.
image

File: src/web/app/src/components/ThemeSelector.tsx (hunk 2 above)
Comment (Genne23v) [reply src/web/app/src/components/ThemeSelector.tsx:2]: This wasn't compiled for me. Somehow it works now.

File: src/web/app/src/components/ThemeSelector.tsx (hunk 3 above)
Comment (Genne23v) [reply src/web/app/src/components/ThemeSelector.tsx:70]: I think you are referring to `console.log`. Let me know if `handleChange` needs to be changed.

Comment (manekenpix) [reply src/web/app/src/components/ThemeSelector.tsx:70]: Yes, I meant only the `console.log`.  :+1:

File: src/web/app/src/components/ThemeSelector.tsx (hunk 4 above)
Comment (Genne23v) [reply src/web/app/src/components/ThemeSelector.tsx:75]: Yeah, it's clearer. Thanks

**Comment by Genne23v** [comment]
I misunderstood your comments. Now it's menu title is not clickable. Thanks for checking!

**Comment by Genne23v** [comment]
@cychu42  I misunderstood your comments. Now it's menu title is not clickable. Thanks for checking!

File: src/web/app/src/components/ThemeSelector.tsx (hunk 2 above)
Comment (Genne23v) [reply src/web/app/src/components/ThemeSelector.tsx:2]: Sorry I think it wasn't a compile issue. Actually it was a lint error. It still gives me `'React' was used before it was defined  no-use-before-define` error.

File: src/web/app/src/components/ThemeSelector.tsx (hunk 4 above)
Comment (Genne23v) [reply src/web/app/src/components/ThemeSelector.tsx:75]: This one also gives me `Redundant double negation               no-extra-boolean-cast` lint error.

File: src/web/app/src/components/ThemeProvider.tsx (hunk 1 above)
Comment (Genne23v) [reply src/web/app/src/components/ThemeProvider.tsx:7]: I removed all possibilities that `preferredTheme` gets undefined, but I have a typescript error in IDE. Since lint is not complaining and the code compile has no problem, I will push without this `undefined`. Thanks for the feedback.

File: src/web/app/src/components/ThemeSelector.tsx (hunk 2 above)
Comment (manekenpix) [reply src/web/app/src/components/ThemeSelector.tsx:2]: My bad, it turned out that my linter wasn't working, but thanks to this I found a bug in our `vscode` settings. You can ignore this comment.

File: src/web/app/src/components/ThemeSelector.tsx (hunk 4 above)
Comment (manekenpix) [reply src/web/app/src/components/ThemeSelector.tsx:75]: [This comment](https://github.com/Seneca-CDOT/telescope/pull/3761/files#r1027118153) also explains what happened here.

**Comment by Genne23v** [comment]
@manekenpix please check my updated branch. I just applied the same colour on selected buttons. Thanks!

**Comment by manekenpix** [comment]
@Genne23v you can click on the  spinning icon to re-request reviews

image

File: src/web/app/src/hooks/use-preferred-theme.ts [hunk 5]
```diff
@@ -1,27 +1,26 @@
...
+import { useLocalStorage } from 'react-use';
 
 /**
  * Combination of localStorage for remembering the user's preference between
  * loads of the app, and initial logic to get the browser's preferred colour.
  */
 export default function usePreferredTheme() {
-  const isDarkThemePreferred = useMedia('(prefers-color-scheme: dark)', false);
-  const [preferredTheme, setPreferredTheme] = useLocalStorage(
+  const [preferredTheme, setPreferredTheme] = useLocalStorage<String>(
```
Comment (manekenpix) [inline src/web/app/src/hooks/use-preferred-theme.ts:9]: I think the issue you're having with the build is related to this. Instead of type `String`, you can create a type that's something like:

```ts
type ThemeType = 'light-default' | 'light-high-contrast' | 'dark-default' | 'dark-dim';
```

Comment (Genne23v) [reply src/web/app/src/hooks/use-preferred-theme.ts:9]: Thanks for great feedback. Let me try again.

Comment (Genne23v) [reply src/web/app/src/hooks/use-preferred-theme.ts]: @manekenpix Sorry that I couldn't check build errors on my local. I have some spaces in my folder name which produces invalid folder error. I made some improvement in type checking and making const variables for repeated use. Please review and let me know if you have any feedback. Thanks

**Review by manekenpix — APPROVED** [review]
Great work. thanks for this feature!

I think there might be a bug with the tooltip, but it's hard for me to reproduce, so let's merge this and we'll file an issue once the bug is clearer.

**Comment by humphd** [comment]
This is really cool

**Comment by Genne23v** [comment]
@manekenpix Thank you so much. I learned a lot with this change. I really appreciate your feedback. I hope someone with better sense of colour updates the palette and add more themes as wished.
